lib/gridfill/_gridfill.pyx

```diff
@@ -0,0 +1,350 @@
+    if int_sum(mask, nlat, nlon) == 0:
+        numiter[0] = 0
+        resmax[0] = 0
+        return
+    # Set initial values for all missing values.
+    initialize_missing(grid, mask, nlat, nlon, initialize_zonal)
+    dp25 = 0.25
+    _numiter = 0
+    _resmax = 0
+    while _numiter < itermax:
+        _resmax = 0
+        _numiter += 1
+        for i in range(nlat):
+            latitude_indices(i, nlat, &im1, &ip1)
+            for j in range(nlon):
+                if mask[i, j]:
+                    longitude_indices(j, nlon, cyclic, &jm1, &jp1)
+                    residual = dp25 * (grid[im1, j] + grid[ip1, j] +
+                                       grid[i, jm1] + grid[i, jp1]) - grid[i, j]
+                    residual *= relaxc
+                    grid[i, j] += residual
+                    _resmax = double_max(double_abs(residual), _resmax)
+        if _resmax <= tolerance:
+            break
+    numiter[0] = _numiter
+    resmax[0] = _resmax
+
+
+def poisson_fill_grids(double[:, :, :] grids,
+                       int[:, :, :] masks,
+                       double relaxc,
+                       double tolerance,
+                       unsigned int itermax,
+                       int cyclic,
+                       int initialize_zonal):
+    """
+    Fill missing values in grids by iterative relaxation.
+
+    **Arguments:**
+
+    * grids [numpy.ndarray[double, ndim=3]]
+       A 3-dimensional `numpy.ndarray` containing grids of values. The
+       grid dimensions are [y, x, n] where n is a non-grid dimension.
+
+    * masks [numpy.ndarray[double, ndim=3]]
+       A 3-dimensional `numpy.ndarray` the same shape as *grids* that
+       contains the grid mask. Valid grid points are indicated with a
+       zero value, and missing values are indicated by a non-zero value.
+
+    * relaxc [double]
+       The relaxation constant, typically 0.45 <= *relaxc* <= 0.6.
+
+    * tolerance [double]
+       Numerical tolerance for convergence of the relaxation scheme.
+       This value is data dependent.
+
+    * itermax [unsigned int]
+       The maximum number of iterations allowed for the relaxation
+       scheme.
+
+    * cyclic [int]
+       If `0` the input grid is assumed not to be cyclic. If non-zero
+       the input grid is assumed to be cyclic in its second dimension.
+
+    * initialize_zonal [int]
+       If non-zero, take the zonal mean as the initial guess for missing
+       values. If zero, use the value `0` as the initial guess for all
+       missing values.
+
+    **Returns:**
+
+    * numiter [numpy.ndarray[unsigned int, ndim=1]
+       The number of iterations used to fill each grid.
+
+    * resmax [numpy.ndarray[double, ndim=1]]
+       The maximum residual value at the end of the iteration for each
+       grid. If this value is larger than the specified *tolerance* then
+       the iteration did not converge.
+
+    """
+    cdef unsigned int grid_num
+    cdef unsigned int nlat = grids.shape[0]
```

Comment:
Should check if `nlat` and `nlon` are at least 3, otherwise the indexing code might wrap an unsigned int and cause a segmentation fault.